File: Manuel/코딩 테스트 교재 내용/011. 스택으로 수열 만들기/1874.cpp
#include <iostream>
#include <vector>
#include <stack>
#include <string>
#define endl "\n"

using namespace std;

int main()
{
    ios_base::sync_with_stdio(false);
    cin.tie(NULL); cout.tie(NULL);

    int N;
    cin >> N;
    vector<int> A(N);
    for(int i = 0; i < N; i++)
    {
        cin >> A[i];
    }

    stack<int> s;
    int num = 1;
    bool result = true;
    string answer = "";

    for(size_t i = 0; i < A.size(); i++)
    {
        int number = A[i];

        if(number >= num)
        {
            while(number >= num)
            {
                s.push(num);
                num++;
                answer += "+\n";
            }
            s.pop();
            answer += "-\n";
        }
        else
        {
            if(s.empty())
            {
                result = false;
                break;
            }
            int n = s.top();
            s.pop();
            if(n > number)
            {
                cout << "NO" << endl;
                result = false;
                break;
            }
            else
            {
                answer += "-\n";
            }
        }
    }
    if(result)
    {
        cout << answer;
    }
    
    return 0;
}

File: Manuel/코딩 테스트 교재 내용/011. 스택으로 수열 만들기/1874.py
N = int(input())
A = [0] * N

for i in range(N):
    A[i] = int(input())

stack = []
num = 1
result = True
answer = ""

for i in range(N):
    number = A[i]
    if number >= num:
        while number >= num:
            stack.append(num)
            num += 1
            answer += "+\n"
        stack.pop()
        answer += "-\n"
    else:
        n = stack.pop()
        if n > number:
            print("NO")
            result = False
            break
        else:
            answer += "-\n"

if result:
    print(answer)